File: src/ultidraft/domain/lexicon.py
"""Per-book pronunciation rules. Applied only to spoken text, never the manuscript."""

from __future__ import annotations

import re
from dataclasses import asdict, dataclass


@dataclass
class LexiconRule:
    written: str
    spoken: str

    def to_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict) -> LexiconRule:
        return cls(
            written=str(data.get("written", "")).strip(),
            spoken=str(data.get("spoken", "")).strip(),
        )


def apply_lexicon(text: str, rules: list[LexiconRule]) -> str:
    spoken = text
    for rule in rules:
        if not rule.written or not rule.spoken:
            continue
        pattern = re.compile(r"\b" + re.escape(rule.written) + r"\b", re.IGNORECASE)
        spoken = pattern.sub(rule.spoken, spoken)
    return spoken

File: android/app/src/main/java/com/ultidraft/domain/Lexicon.kt
package com.ultidraft.domain

/** Per-book pronunciation rules. Applied only to spoken text, never the manuscript. */
data class LexiconRule(val written: String, val spoken: String) {
    fun toJson(): JsonValue = JsonValue.obj(
        "written" to JsonValue.str(written),
        "spoken" to JsonValue.str(spoken),
    )

    companion object {
        fun fromJson(value: JsonValue): LexiconRule = LexiconRule(
            written = value.string("written").trim(),
            spoken = value.string("spoken").trim(),
        )
    }
}

fun applyLexicon(text: String, rules: List<LexiconRule>): String {
    var spoken = text
    for (rule in rules) {
        if (rule.written.isEmpty() || rule.spoken.isEmpty()) continue
        val pattern = Regex("\\b" + Regex.escape(rule.written) + "\\b", RegexOption.IGNORE_CASE)
        // Lambda form: the result is inserted literally, so "$" in a rule stays safe.
        spoken = pattern.replace(spoken) { rule.spoken }
    }
    return spoken
}
